usr.bin/who/who.1
.\"	$NetBSD: who.1,v 1.26 2021/02/23 07:15:41 wiz Exp $
.\"
.\" Copyright (c) 1986, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)who.1	8.2 (Berkeley) 12/30/93
.\"
.Dd September 1, 2019
.Dt WHO 1
.Os
.Sh NAME
.Nm who
.Nd display who is logged in
.Sh SYNOPSIS
.Nm
.Op Fl abdHlmqrsTtuv
.Op Ar file
.Nm
.Ar am i
.Sh DESCRIPTION
The
.Nm
utility displays a list of all users currently logged on, showing for
each user the login name, tty name, the date and time of login, and
hostname if not local.
.Pp
Available options:
.Bl -tag -width file
.It Fl a
Same as
.Fl bdlprTtuv .
.It Fl b
Time of last system boot.
.It Fl d
Print dead processes.
.It Fl H
Write column headings above the regular output.
.It Fl l
Print system login processes.
.It Fl m
Only print information about the current terminal.
This is the POSIX way of saying
.Nm
.Ar am i .
.It Fl p
Print active processes spawned by
.Xr init 8 .
.It Fl q
.Dq Quick mode :
List only the names and the number of users currently logged on.
When this option is used, all other options are ignored.
.It Fl r
Print the current runlevel.
Supported runlevels are:
.Bl -tag -width "s (SINGLE_USER)"
.It Dv d Pq Dv DEATH
The system has halted.
.It Dv s Pq Dv SINGLE_USER
The system is running in single user mode.
.It Dv r Pq Dv RUNCOM
The system is executing
.Pa /etc/rc .
.It Dv t Pq Dv READ_TTYS
The system is processing
.Pa /etc/ttys .
.It Dv m Pq Dv MULTI_USER
The system is running in multi-user mode.
.It Dv T Pq Dv CLEAN_TTYS
The system is in the process of stopping processes
associated with terminal devices.
.It Dv c Pq Dv CATATONIA
The system is in the process of shutting down and will
not create new processes.
.El
.It Fl s
List only the name, line and time fields.
This is the default.
.It Fl T
Print a character after the user name indicating the state of the
terminal line:
.Sq +
if the terminal is writable;
.Sq -
if it is not;
and
.Sq \&?
if a bad line is encountered.
.It Fl t
Print last system clock change.
.It Fl u
Print the idle time for each user, and the associated process ID.
.It Fl v
When printing of more information is requested with
.Fl u ,
this switch can be used to also printed
process termination signals,
process exit status,
session id for windowing
and the type of the entry, see documentation of ut_type in
.Xr getutxent 3 .
.It Ar \&am I
Returns the invoker's real user name.
.It Ar file
By default,
.Nm
gathers information from the file
.Pa /var/run/utmpx .
An alternative
.Ar file
may be specified which is usually
.Pa /var/log/wtmpx
(or
.Pa /var/log/wtmp ,
or
.Pa /var/log/wtmpx.[0-6]
or
.Pa /var/log/wtmp.[0-6]
depending on site policy as
.Pa wtmpx
can grow quite large and daily versions may or may not
be kept around after compression by
.Xr ac 8 ) .
The
.Pa wtmpx
and
.Pa wtmp
file contains a record of every login, logout,
crash, shutdown and date change
since
.Pa wtmpx
and
.Pa wtmp
were last truncated or
created.
.El
.Pp
If
.Pa /var/log/wtmpx
or
.Pa /var/log/wtmp
are being used as the file, the user name may be empty
or one of the special characters '|', '}' and '~'.
Logouts produce an output line without any user name.
For more information on the
special characters, see
.Xr utmp 5 .
.Sh FILES
.Bl -tag -width /var/log/wtmp.[0-6] -compact
.It Pa /var/run/utmp
.It Pa /var/run/utmpx
.It Pa /var/log/wtmp
.It Pa /var/log/wtmp.[0-6]
.It Pa /var/log/wtmpx
.It Pa /var/log/wtmpx.[0-6]
.El
.Sh SEE ALSO
.Xr last 1 ,
.Xr mesg 1 ,
.Xr users 1 ,
.Xr getuid 2 ,
.Xr utmp 5 ,
.Xr utmpx 5
.Sh STANDARDS
The
.Nm
utility is expected to conform to
.St -p1003.2-92 .
.Sh HISTORY
A
.Nm
utility appeared in
.At v1 :
.Lk https://www.bell-labs.com/usr/dmr/www/man14.pdf


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2784 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 9.0 to NetBSD 10.0:
	openldap: Import 2.4.48. [christos 20190808]
	boot(8): GPT and RAIDframe support for x86 bootstrap [manu 20190818]
	dhcpcd(8): Import dhcpcd-8.0.3 [roy 20190821]
	inet6: Fix sending RTM_MISS for unresolvable addresses and restore
		ND6_LLINFO_WAITDELETE so that any send at this point can
		return EHOSTDOWN to match inet behaviour. [roy 20190901]
	inet: Fix sending RTM_MISS for unresolvable addresses. [roy 20190901]
	dhcpcd(8): Import dhcpcd-8.0.4 [roy 20190904]
	bind: Import version 9.14.5. [christos 20190905]
	resolvconf(8): Import openresolv-3.9.2 [roy 20190908]
	boot(8): multiboot 2 support, Xen can now boot from EFI [manu 20190913]
	dhcpcd(8): Import dhcpcd-8.0.6 [roy 20190913]
	kernel: Add vHCI, a driver that allows to send and receive USB
		packets from userland. [maxv 20190914]
	mos(4): Ported driver for MosChip MCS7730/7830/7832 USB ethernet.
		[mrg 20190920]
	gcc(1): Import GCC 8.3. [mrg 20191001]
	libpcap: Import 1.9.1. [christos 20191001]
	tcpdump(8): Import 4.9.3. [christos 20191001]
	rescue(8): Moved into a separate, rescue-only set. [maya 20191001]
	byacc: Update to 20190617. [christos 20191006]
	tzdata updates from 2019a to 2019c (incl 2019b) [kre 20191008]
	services(5): Update to version 2019-10-04 from IANA.
		[sevan 20191009]
	dhcpcd(8): Import dhcpcd-8.1.0 [roy 20191011]
	OpenSSH: Import 8.1. [christos 20191012]
	acpi(4): Updated ACPICA to 20190816. [christos 20191015]
	sup: removed, use ftp://ftp.astron.com/pub/sup [christos 20191015]
	dhcpcd(8): Import dhcpcd-8.1.1 [roy 20191016]
	evbarm: Add support for the NanoPi R1 and Duo2. [bad 20191016]
	bind: Import version 9.14.7. [christos 20191017]
	libc: Changed the iconv(3) function signature to the POSIX variation
		[kamil 20191024]
	nct(4): Add Nuvoton NCT5104D GPIO driver. [ad 20191025]
	vio9p(4): Add virtio-9p front-end driver. [ozaki-r 20191028]
	mount_9p(8): Enable to mount a filesystem exported via virtio-9p.
		[ozaki-r 20191028]
	evbarm: Update TI AM335x SoC support to use FDT based configuration.
		[jmcneill 20191028]
	evbarm: Update TI OMAP3 SoC support to use FDT based configuration.
		[jmcneill 20191101]
	amd64: Add support for kCSan - Kernel Concurrency Sanitizer.
		[maxv 20191105]
	tmux(1): Imported 2.9a. [christos 20191112]
	dhcpcd(8): Import 8.1.2. [roy 20191113]
	amd64: Add support for kMSan - Kernel Memory Sanitizer.
		[maxv 20191114]
	qat(4): Add opencrypto driver for Intel QuickAssist.
		Atom C2XXX, C3XXX, Xeon D-21XX and D-15XX, C62X chipsets and
		QuickAssist Adapter 8960/8970.
		[hikaru 20191120]
	select(2), poll(2): Performance and scalability improvements
		[ad 20191121]
	amd64, i386: Performance and scalability improvements for TLB
		shootdown [ad 20191121]
	bind: Import version 9.14.8. [christos 20191127]
	terminfo: Import 20190609 [christos 20191127]
	onewire(4), owtemp(4): Improve reliability and reduce CPU overhead
		slightly. [ad 20191130]
	scheduler: Improve performance and interactivity. [ad 20191201]
	vfs: Reduce false sharing and lock overhead during normal operation.
		[ad 20191201]
	evbarm: Added driver for Allwinner Crypto Engine random number
		generator [riastradh 20191209]
	ixl(4): Ported driver for Intel Ethernet 700 series
		[yamaguchi 20191210]
	acpi(4): Updated ACPICA to 20191213. [christos 20191214]
	uvm: Replace red-black tree page lookup with radix tree lookup,
		including tagging and gang lookup of pages.  Original work
		done by yamt@. [ad 20191214]
	nsd: Import 4.2.4. [christos 20191215]
	unbound(8): Import 1.9.6. [christos 20191215]
	openpam(3): update to 20190224 (tabebuia) [christos 20191215]
 	kerberos(8): Update to Heimdal-7.7.0 [christos 20191215]
	file(1): Upgraded to 5.38. [christos 20191216]
	uvm: Replace global counters with per-CPU counters, significantly
		reducing system time on workloads that are VM-system heavy.
		[ad 20191216]
	anxdp(4): Add driver for Analogix DisplayPort core found in
		Rockchip RK3399. [jakllsch 20191219]
	net80211: Require SSID configuration to associate with an open AP.
		[jakllsch 20191219]
	dhcpcd(8): Import dhcpcd-8.1.4 [roy 20191220]
	uvm: Rewrite the page allocator to be more efficient and somewhat CPU
		topology aware.  Add preliminary NUMA support.  [ad 20191227]
	crunchgen(1): Added the ability to pass variables in the submake
		command line, and removed all command like flags that are
		not needed anymore [christos 20191229]
	uvm: Reduce lock contention by making the maintentance of page
		replacement state more concurrent. [ad 20191231]
	aq(4): Add Aquantia 10G network adapter driver [ryo 20200101]
	dhcpcd(8): Import dhcpcd-8.1.5 [roy 20200103]
	tmux(1): Imported 3.0a. [christos 20200106]
	amd64, i386: Replace the global PV hash table with radix trees,
		yielding significant performance and scalability benefits.
		[ad 20200112]
	scheduler: Make the scheduler topology aware.  It understands and
		makes good use of HyperThreading/SMT, spreads the load evenly

usr.bin/who/who.c
/*	$NetBSD: who.c,v 1.25 2015/11/21 15:01:43 christos Exp $	*/

/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Michael Fischbein.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1989, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)who.c	8.1 (Berkeley) 6/6/93";
#endif
__RCSID("$NetBSD: who.c,v 1.25 2015/11/21 15:01:43 christos Exp $");
#endif /* not lint */

#include <sys/types.h>
#include <sys/stat.h>

#include <err.h>
#include <locale.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#ifdef SUPPORT_UTMP
#include <utmp.h>
#endif
#ifdef SUPPORT_UTMPX
#include <utmpx.h>
#endif

#include "utmpentry.h"

static void output_labels(void);
static void who_am_i(const char *, int);
static void usage(void) __dead;
static void process(const char *, int);
static void eprint(const struct utmpentry *);
static void print(const char *, const char *, time_t, const char *, pid_t pid,
    uint16_t term, uint16_t xit, uint16_t sess, uint16_t type);
static void quick(const char *);

static int show_term;			/* show term state */
static int show_idle;			/* show idle time */
static int show_details;		/* show exit status etc. */

struct ut_type_names {
  int type;
  const char *name;
} ut_type_names[] = {
#ifdef SUPPORT_UTMPX
  { EMPTY, "empty" }, 
  { RUN_LVL, "run level" }, 
  { BOOT_TIME, "boot time" }, 
  { OLD_TIME, "old time" }, 
  { NEW_TIME, "new time" }, 
  { INIT_PROCESS, "init process" }, 
  { LOGIN_PROCESS, "login process" }, 
  { USER_PROCESS, "user process" }, 
  { DEAD_PROCESS, "dead process" }, 
#if defined(_NETBSD_SOURCE)
  { ACCOUNTING, "accounting" }, 
  { SIGNATURE, "signature" },
  { DOWN_TIME, "down time" },
#endif /* _NETBSD_SOURCE */
#endif /* SUPPORT_UTMPX */
  { -1, "unknown" }
};

int
main(int argc, char *argv[])
{
	int c, only_current_term, show_labels, quick_mode, default_mode;
	int et = 0;

	setlocale(LC_ALL, "");

	only_current_term = show_term = show_idle = show_labels = 0;
	quick_mode = default_mode = 0;

	while ((c = getopt(argc, argv, "abdHlmpqrsTtuv")) != -1) {
		switch (c) {
		case 'a':
			et = -1;
			show_idle = show_details = 1;
			break;
		case 'b':
			et |= (1 << BOOT_TIME);
			break;
		case 'd':
			et |= (1 << DEAD_PROCESS);
			break;
		case 'H':
			show_labels = 1;
			break;
		case 'l':
			et |= (1 << LOGIN_PROCESS);
			break;
		case 'm':
			only_current_term = 1;
			break;
		case 'p':
			et |= (1 << INIT_PROCESS);
			break;
		case 'q':
			quick_mode = 1;
			break;
		case 'r':
			et |= (1 << RUN_LVL);
			break;
		case 's':
			default_mode = 1;
			break;
		case 'T':
			show_term = 1;
			break;
		case 't':
			et |= (1 << NEW_TIME);
			break;
		case 'u':
			show_idle = 1;
			break;
		case 'v':
			show_details = 1;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	if (et != 0)
		etype = et;

	if (chdir("/dev")) {
		err(EXIT_FAILURE, "cannot change directory to /dev");
		/* NOTREACHED */
	}

	if (default_mode)
		only_current_term = show_term = show_idle = 0;

	switch (argc) {
	case 0:					/* who */
		if (quick_mode) {
			quick(NULL);
		} else if (only_current_term) {
			who_am_i(NULL, show_labels);
		} else {
			process(NULL, show_labels);
		}
		break;
	case 1:					/* who utmp_file */
		if (quick_mode) {
			quick(*argv);
		} else if (only_current_term) {
			who_am_i(*argv, show_labels);
		} else {
			process(*argv, show_labels);
		}
		break;
	case 2:					/* who am i */
		who_am_i(NULL, show_labels);
		break;
	default:
		usage();
		/* NOTREACHED */
	}

	return 0;
}

static char *
strrstr(const char *str, const char *pat)
{
	const char *estr;
	size_t len;
	if (*pat == '\0')
		return __UNCONST(str);

	len = strlen(pat);

	for (estr = str + strlen(str); str < estr; estr--)
		if (strncmp(estr, pat, len) == 0)
			return __UNCONST(estr);
	return NULL;
}

static void
who_am_i(const char *fname, int show_labels)
{
	struct passwd *pw;
	const char *p;
	char *t;
	time_t now;
	struct utmpentry *ehead, *ep;

	/* search through the utmp and find an entry for this tty */
	if ((p = ttyname(STDIN_FILENO)) != NULL) {

		/* strip directory prefixes for ttys */
		if ((t = strrstr(p, "/pts/")) != NULL ||
		    (t = strrchr(p, '/')) != NULL)
			p = t + 1;

		(void)getutentries(fname, &ehead);
		for (ep = ehead; ep; ep = ep->next)
			if (strcmp(ep->line, p) == 0) {
				if (show_labels)
					output_labels();
				eprint(ep);
				return;
			}
	} else
		p = "tty??";

	(void)time(&now);
	pw = getpwuid(getuid());
	if (show_labels)
		output_labels();
	print(pw ? pw->pw_name : "?", p, now, "", getpid(), 0, 0, 0, 0);
}

static void
process(const char *fname, int show_labels)
{
	struct utmpentry *ehead, *ep;
	(void)getutentries(fname, &ehead);
	if (show_labels)
		output_labels();
	for (ep = ehead; ep != NULL; ep = ep->next)
		eprint(ep);
}

static void
eprint(const struct utmpentry *ep)
{
	print(ep->name, ep->line, (time_t)ep->tv.tv_sec, ep->host, ep->pid,
	    ep->term, ep->exit, ep->sess, ep->type);
}

static void
print(const char *name, const char *line, time_t t, const char *host,
    pid_t pid, uint16_t term, uint16_t xit, uint16_t sess, uint16_t type)
{
	struct stat sb;
	char state;
	static time_t now = 0;
	time_t idle;
	const char *types = NULL;
	size_t i;
	char *tstr;

	state = '?';
	idle = 0;

	for (i = 0; ut_type_names[i].type >= 0; i++) {
		types = ut_type_names[i].name;
		if (ut_type_names[i].type == type)
			break;
	}
	
	if (show_term || show_idle) {
		if (now == 0)
			time(&now);
		
		if (stat(line, &sb) == 0) {
			state = (sb.st_mode & 020) ? '+' : '-';
			idle = now - sb.st_atime;
		}
		
	}

	(void)printf("%-*.*s ", (int)maxname, (int)maxname, name);

	if (show_term)
		(void)printf("%c ", state);

	(void)printf("%-*.*s ", (int)maxline, (int)maxline, line);
	tstr = ctime(&t);
	(void)printf("%.12s ", tstr ? tstr + 4 : "?");

	if (show_idle) {
		if (idle < 60) 
			(void)printf("  .   ");
		else if (idle < (24 * 60 * 60))
			(void)printf("%02ld:%02ld ", 
				     (long)(idle / (60 * 60)),
				     (long)(idle % (60 * 60)) / 60);
		else
			(void)printf(" old  ");

		(void)printf("\t%6d", pid);
		
		if (show_details) {
			if (type == RUN_LVL)
				(void)printf("\tnew=%c old=%c", term, xit);
			else
				(void)printf("\tterm=%d exit=%d", term, xit);
			(void)printf(" sess=%d", sess);
			(void)printf(" type=%s ", types);
		}
	}
	
	if (*host)
		(void)printf("\t(%.*s)", (int)maxhost, host);
	(void)putchar('\n');
}

static void
output_labels(void)
{
	(void)printf("%-*.*s ", (int)maxname, (int)maxname, "USER");

	if (show_term)
		(void)printf("S ");
	
	(void)printf("%-*.*s ", (int)maxline, (int)maxline, "LINE");
	(void)printf("WHEN         ");

	if (show_idle) {
		(void)printf("IDLE  ");
		(void)printf("\t   PID");
	
		(void)printf("\tCOMMENT");
	}		

	(void)putchar('\n');
}

static void
quick(const char *fname)
{
	struct utmpentry *ehead, *ep;
	int num = 0;

	(void)getutentries(fname, &ehead);
	for (ep = ehead; ep != NULL; ep = ep->next) {
		(void)printf("%-*s ", (int)maxname, ep->name);
		if ((++num % 8) == 0)
			(void)putchar('\n');
	}
	if (num % 8)
		(void)putchar('\n');

	(void)printf("# users = %d\n", num);
}

static void
usage(void)
{
	(void)fprintf(stderr, "Usage: %s [-abdHlmqrsTtuv] [file]\n\t%s am i\n",
	    getprogname(), getprogname());
	exit(EXIT_FAILURE);
}
